### ISSUE TITLE
db: block property filtering does not respect range tombstones

### ISSUE THREAD

**Issue by erikgrinaker** [cockroachdb/pebble#1786]
I've seen both point and range keys reappear after writing a Pebble range tombstone across them, over in https://github.com/cockroachdb/cockroach/pull/82667/files#diff-8c73ba608e99d883eaaa8779aa0b2ee85808bcc4019d97eec9dc577fa35e1ce9R872-R873. I first flush an SST with point keys, then flush an SST with range keys, and then flush an SST with Pebble range tombstones across them. It appears to be racy, possibly with flushes or compactions.

**Comment by erikgrinaker** [comment]
It appears that, for range keys, `InternalKeyKindRangeKeyDelete` is exposed to the block property collector, so we can detect it and record it as spanning all timestamps. However, for point keys, we only expose `InternalKeyKindRangeDelete` to table property collectors, not block property collectors:

https://github.com/cockroachdb/pebble/blob/71d17c2a007bfad5111a229ba325d30251b88a41/sstable/writer.go#L813-L818

Would the preferred solution here be to expose `InternalKeyKindRangeDelete` to block property collectors as well, and then record `MinTimestamp` to `MaxTimestamp` for any block that contains a range tombstone of either kind (point key or range key)?

I also think we may have a bug here in existing CRDB releases, since it doesn't appear like the table property collector will do anything special with these, and simply ignore them since the key bounds don't have timestamps.

**Comment by jbowens** [comment]
> It's unclear whether this is a bug or working as intended,

I discovered this as well in #1725. My understanding from talking with @sumeerbhola is that this was intended, and the motivation for the dual-iterator approach of CockroachDB time-bound iterators. Currently, we're only guaranteed that a Pebble iterator configured with the MVCC filtering will surface all keys within the configured timestamp range. It doesn't guarantee that it won't surface keys outside of the timestamp range, or (surprisingly) that it won't surface deleted keys.

That said, I definitely intend to tighten these semantics, because we need tighter semantics in order to use block-property filtering to implement efficient range-key masking. I've got a wip in #1737, but it still needs some work to clean it up.

**Comment by erikgrinaker** [comment]
> It doesn't guarantee that it won't surface keys outside of the timestamp range, or (surprisingly) that it won't surface deleted keys.

Yeah, the latter is definitely a footgun. For example, I originally made use of TBIs for MVCC range tombstone conflict checks (which only needed to look for future keys within the span), which could have given false conflicts if Pebble range tombstones were ignored. We'll likely do something similar to check for newer writes when GCing MVCC range tombstones.

>I definitely intend to tighten these semantics, because we need tighter semantics in order to use block-property filtering to implement efficient range-key masking. I've got a wip in https://github.com/cockroachdb/pebble/pull/1737, but it still needs some work to clean it up.

Nice, looks like we won't need to handle these in CRDB property filters then. Will the same logic apply to `InternalKeyKindRangeKeyDelete`, or will we have to explicitly handle those?

**Comment by jbowens** [comment]
> Will the same logic apply to InternalKeyKindRangeKeyDelete, or will we have to explicitly handle those?

Same logic will apply for them 👍

**Comment by erikgrinaker** [comment]
I found that we have a similar problem with `RangeKeyUnset` and and `Delete`: if we have keys (either points or ranges) at e.g. `@1` and `@2` in one SST, and then clear the `@2` key in a different SST, filtering by `@1` will reveal the key at `@2` because the SST with the clear does not satisfy the filter. Come to think of it, this would also apply when overwriting keys, where an old version of a key could be revealed by a property filter.

It seems to me that the only safe way to use these would be to prune values outside of the timestamp bounds, to ensure that we see all relevant updates to all exposed keys.

**Comment by erikgrinaker** [comment]
Submitted a PR to call out these caveats, since they aren't currently documented: https://github.com/cockroachdb/cockroach/pull/83385

**Comment by jbowens** [comment]
Yeah, the `RangeKeyDelete` and `RangeDelete` case can cause deleted keys within the timestamp bounds to be surfaced, whereas `RangeKeyUnset` and `Delete` can cause deleted keys outside the timestamp bounds to be surfaced.

> It seems to me that the only safe way to use these would be to prune/ignore values outside of the timestamp bounds, to ensure that we see all relevant updates to all surfaced keys.

It would be great if `pkg/storage` only exposed an iterator that would automatically perform that pruning, so that iteration results are deterministic and there's no risk of misuse by a consumer. I'm not sure if that would impact performance or require substantial refactoring.
